lm/lm_image.py

```diff
@@ -0,0 +1,120 @@
+        renormalize=False,
+    ):
+        kwargs = locals()
+        del kwargs["self"]
+
+        self.text_file = text_file
+        self.renormalize = renormalize
+
+        self.config, self.post_config, self.num_images = LmImageJob.create_config(
+            **kwargs
+        )
+        self.exe = self.select_exe(crp.lm_util_exe, "lm-util")
+
+        self.log_file = self.log_file_output_path("lm_image", crp, False)
+        self.lm_images = {
+            i: self.output_path(f"lm-{i}.image", cached=True)
+            for i in range(1, self.num_images + 1)
+        }
+
+        self.rqmt = {"time": 1, "cpu": 1, "mem": 2}
+
+    def tasks(self):
+        yield Task("create_files", resume="create_files", mini_task=True)
+        yield Task("run", resume="run", rqmt=self.rqmt)
+
+    def create_files(self):
+        self.write_config(self.config, self.post_config, "lm_image.config")
+        extra_code = (
+            ":${{THEANO_FLAGS:="
+            "}}\n"
+            'export THEANO_FLAGS="$THEANO_FLAGS,device={0},force_device=True"\n'
+            'export TF_DEVICE="{0}"'.format(
+                "gpu" if self.rqmt.get("gpu", 0) > 0 else "cpu"
+            )
+        )
+        self.write_run_script(self.exe, "lm_image.config", extra_code=extra_code)
+
+    def run(self):
+        self.run_script(1, self.log_file)
+        for i in range(1, self.num_images + 1):
+            shutil.move(f"lm-{i}.image", self.lm_images[i].get_path())
+
+    def cleanup_before_run(self, cmd, retry, *args):
+        util.backup_if_exists("lm_image.log")
+
+    @classmethod
+    def find_arpa_lms(cls, lm_config, lm_post_config=None):
+        result = []
+
+        def has_image(c, pc):
+            res = c._get("image") is not None
+            res = res or (pc is not None and pc._get("image") is not None)
+            return res
+
+        if lm_config.type == "ARPA":
+            if not has_image(lm_config, lm_post_config):
+                result.append((lm_config, lm_post_config))
+        elif lm_config.type == "combine":
+            for i in range(1, lm_config.num_lms + 1):
+                sub_lm_config = lm_config["lm-%d" % i]
+                sub_lm_post_config = (
+                    lm_post_config["lm-%d" % i] if lm_post_config is not None else None
+                )
+                result += cls.find_arpa_lms(sub_lm_config, sub_lm_post_config)
+        return result
+
+    @classmethod
+    def create_config(
+        cls,
+        crp,
+        extra_config,
+        extra_post_config,
+        encoding,
+        renormalize,
+        **kwargs,
+    ):
+        config, post_config = rasr.build_config_from_mapping(
+            crp, {"lexicon": "lm-util.lexicon", "language_model": "lm-util.lm"}
+        )
+        del (
+            config.lm_util.lm.scale
+        )  # scale not considered here, delete to remove ambiguity
+
+        config.lm_util.action = "load-lm"
+        config.lm_util.file = crp.language_model_config.file
```

Comment:
This is probably due to the RASR implementation of lm-util, but why the explicit `config.lm_util.file` and the implicit (via `rasr.build_config_from_mapping` and `crp.language_model_config`) `config.lm_util.lm.file`? What happens if there are multiple LMs defined (e.g. combined LM with sub-LMs) what if the LM is not in ARPA format and does not have a `.file`?

Reply:
I was not sure about this part, I thought `file` was a necessary parameter for the `lm_util` job, so I added it ... It seemed weird to me too...

Reply:
@michelwi Is this `file` parameter not a necessary one for `lm_util` in `RASR`?

Reply:
I do not know as I never used it. From https://github.com/rwth-i6/rasr/blob/843d1eb5f2137b57dcc6cf4bc9dce44e85d66f27/src/Tools/Lm/LmUtilityTool.cc#L93 as well as Lines 132 and following, I gather that the `config.lm_util.file` is a plain text file, that is used to compute the perplexities of the given LM on.

Also from https://github.com/rwth-i6/rasr/blob/843d1eb5f2137b57dcc6cf4bc9dce44e85d66f27/src/Tools/Lm/LmUtilityTool.cc#L119-L123 I don't see the action `"load-lm"` defined. Did/Will you/someone implement this functionality?

Reply:
maybe @curufinwe who wrote the lm-util tool can help clarify.

Reply:
I am using RASR in `/home/ebeck/dev/full_rasr/rasr/` and there we can find a `load-lm` at line 91 in `/home/ebeck/dev/full_rasr/rasr/src/Tools/Lm/LmUtilityTool.cc`. Maybe the changes haven't been pushed yet?

Reply:
no, the option does exist in the closed source repo of AppTek. If you want to use it here, we should ask permission to port it to the public repo. @curufinwe